### PR TITLE
Feature/issue 13561 2.3

### PR DESCRIPTION
**Pull request by bnymn**
The original PR is #14660.

Enable Magento 2 to connect MySQL through SSL.

### Description
We have made some modification to enable Magento 2 connect MySQL through SSL.

During the installation, there was no option to force MySQL connection through SSL. With this PR, the following options will be enabled for you to use SSL.
```
bin/magento setup:install \
	--db-host=<MYSQL_HOST> \
	--db-name=<MYSQL_DATABASE> \
	--db-user=<MYSQL_USERNAME> \
	--db-password=<MYSQL_PASSWORD> \
	--db-ssl-key=<MYSQL_CLIENT_KEY> \
	--db-ssl-cert=<MYSQL_CLIENT_CERT> \
	--db-ssl-ca=<MYSQL_SERVER_CERT>
```
### Fixed Issues (if relevant)
1. magento/magento2#13561: Magento 2 and SSL connection to MySQL

### Manual testing scenarios
There are two scenarios to test this feature. First one is by installing Magento via command-line, and the second one is using the Magento Setup UI via browser.

#### 1. Installing via Command-line
1. Install a MySQL server which accepts connections through SSL.
2. Copy these files into a folder which is accessible by Magento: `client-key.pem`, `client-cert.pem`, `ca.pem`.
3. Install a fresh Magento with the following command-line command:
```
bin/magento setup:install \
	--db-host=<MYSQL_HOST> \
	--db-name=<MYSQL_DATABASE> \
	--db-user=<MYSQL_USERNAME> \
	--db-password=<MYSQL_PASSWORD> \
	--db-ssl-key=<PATH/TO/CLIENT-KEY.pem> \
	--db-ssl-cert=<PATH/TO/CLIENT-CERT.pem> \
	--db-ssl-ca=<PATH/TO/CA.pem>
```
4. Observe that `driver_options` is added into `env.php` file.

#### 2. Installing via Magento Setup UI
1. Install a MySQL server which accepts connections through SSL.
2. Copy these files into a folder which is accessible by Magento: `client-key.pem`, `client-cert.pem`, `ca.pem`.
3. Install a fresh Magento via Setup UI.
screencapture-magento2-local-setup-2018-09-16-10_04_56
4. Observe that `driver_options` is added into `env.php` file.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=18075) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @bnymn. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by bnymn** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @bnymn. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @bnymn, here is your [new Magento instance](https://pr-18075.engcom.dev.magento.com).
Admin access: [https://pr-18075.engcom.dev.magento.com/admin](https://pr-18075.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`

**Comment by bnymn** [comment]
Hi @okorshenko , @ishakhsuvarov 

I am getting the following error during code sniff. I could not resolve this issue, since I don't know where the problem is. Would you please help me solve the problem?

```
1) Magento\Test\Php\LiveCodeTest::testCodeStyle
TypeError: Argument 2 passed to Magento\Sniffs\Annotation\ClassAnnotationStructureSniff::validateAnnotationBlockExists() must be of the type integer, boolean given, called in /home/travis/build/magento/magento2/dev/tests/static/framework/Magento/Sniffs/Annotation/ClassAnnotationStructureSniff.php on line 100
/home/travis/build/magento/magento2/dev/tests/static/framework/Magento/Sniffs/Annotation/ClassAnnotationStructureSniff.php:76
/home/travis/build/magento/magento2/dev/tests/static/framework/Magento/Sniffs/Annotation/ClassAnnotationStructureSniff.php:100
/home/travis/build/magento/magento2/vendor/squizlabs/php_codesniffer/src/Files/File.php:497
/home/travis/build/magento/magento2/vendor/squizlabs/php_codesniffer/src/Files/LocalFile.php:91
/home/travis/build/magento/magento2/vendor/squizlabs/php_codesniffer/src/Runner.php:585
/home/travis/build/magento/magento2/vendor/squizlabs/php_codesniffer/src/Runner.php:394
/home/travis/build/magento/magento2/vendor/squizlabs/php_codesniffer/src/Runner.php:114
/home/travis/build/magento/magento2/dev/tests/static/framework/Magento/TestFramework/CodingStandard/Tool/CodeSniffer.php:105
/home/travis/build/magento/magento2/dev/tests/static/testsuite/Magento/Test/Php/LiveCodeTest.php:270
```

File: setup/src/Magento/Setup/Controller/DatabaseCheck.php [hunk 1]
```diff
@@ -37,12 +38,43 @@ public function indexAction()
...
+            );
             $tablePrefix = isset($params['tablePrefix']) ? $params['tablePrefix'] : '';
             $this->dbValidator->checkDatabaseTablePrefix($tablePrefix);
             return new JsonModel(['success' => true]);
         } catch (\Exception $e) {
             return new JsonModel(['success' => false, 'error' => $e->getMessage()]);
         }
     }
+
+    /**
```
Comment (slavvka) [inline setup/src/Magento/Setup/Controller/DatabaseCheck.php:71]: Please add a description according to https://devdocs.magento.com/guides/v2.3/coding-standards/docblock-standard-general.html#inheritdoc

**Review by slavvka — CHANGES_REQUESTED** [review]

**Comment by slavvka** [comment]
Hey @bnymn as you already know there were introduced the new DocBlock sniffer. Could you please consider fixing static tests according to the guide https://devdocs.magento.com/guides/v2.3/coding-standards/docblock-standard-general.html#inheritdoc?
Thank you

**Comment by bnymn** [comment]
Hi @slavvka ,

There are a lot of **line exceeds** warnings in the PR Quality. Should I fix those?

**Comment by slavvka** [comment]
> There are a lot of **line exceeds** warnings in the PR Quality. Should I fix those?

No, you may ignore Codacy

**Comment by bnymn** [comment]
Hi @slavvka ,

Does this PR needs update?

**Comment by p-bystritsky** [comment]
We need to make sure constant /PDO:MYSQL_ATTR_SSL_VERIFY_SERVER_CERT have the same value for any PHP version > 7.1.3

File: vendor/.htaccess [hunk 2]
```diff
@@ -1,8 +0,0 @@
-<IfVersion < 2.4>
```
Comment (sivaschenko) [inline vendor/.htaccess]: Please revert the `vendor/.htaccess` removal

File: setup/src/Magento/Setup/Validator/AdminCredentialsValidator.php [hunk 3]
```diff
@@ -29,21 +30,29 @@ class AdminCredentialsValidator
...
      * @param \Magento\Setup\Module\ConnectionFactory $connectionFactory
      * @param \Magento\Setup\Module\SetupFactory $setupFactory
+     * @param DriverOptions|null $driverOptions
      */
     public function __construct(
         \Magento\Setup\Model\AdminAccountFactory $adminAccountFactory,
         \Magento\Setup\Module\ConnectionFactory $connectionFactory,
-        \Magento\Setup\Module\SetupFactory $setupFactory
+        \Magento\Setup\Module\SetupFactory $setupFactory,
+        DriverOptions $driverOptions
```
Comment (sivaschenko) [inline setup/src/Magento/Setup/Validator/AdminCredentialsValidator.php]: Please add a new dependency in a backward compatible way: last optional parameter, the property should be initialized using object manager if it was not passed to the constructor

File: setup/src/Magento/Setup/Model/ConfigOptionsList/DriverOptions.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        foreach ($driverOptionKeys as $configKey => $driverOptionKey) {
+            if ($this->optionExists($options, $driverOptionKey)) {
+                $driverOptions[$configKey] = $options[$driverOptionKey];
+            }
+        }
+        foreach ($booleanDriverOptionKeys as $configKey => $driverOptionKey) {
+            $driverOptions[$configKey] = $this->booleanValue($options, $driverOptionKey);
+        }
+
+        return $driverOptions;
```
Comment (sivaschenko) [inline setup/src/Magento/Setup/Model/ConfigOptionsList/DriverOptions.php:43]: I think it would be good to provide caching for `$driverOptions`

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @bnymn , thanks for the pull request!

Please see my review comments

**Comment by sivaschenko** [comment]
@buskamuza can you please take a look at this pull request and SVC failures and let us know it can be merged to 2.3-develop branch.

File: setup/src/Magento/Setup/Validator/AdminCredentialsValidator.php (hunk 3 above)
Comment (buskamuza) [reply setup/src/Magento/Setup/Validator/AdminCredentialsValidator.php]: It also should have default value `null` make it optional.

**Review by buskamuza — CHANGES_REQUESTED** [review]

**Comment by buskamuza** [comment]
If it is absolutely necessary for 2.3.x patch release, I can support it. Adding constants doesn't look a big risk. But it is a feature change (improvement). Please work with PO in the first place.
The procedure for breaking changes approval requires multiple people to be involved. An architect, PO and release manager at the first level.

File: setup/src/Magento/Setup/Validator/AdminCredentialsValidator.php (hunk 3 above)
Comment (bnymn) [reply setup/src/Magento/Setup/Validator/AdminCredentialsValidator.php]: I have just sent a commit to make it optional.

File: vendor/.htaccess (hunk 2 above)
Comment (bnymn) [reply vendor/.htaccess]: I don't remember deleting this file. It must have been an accident. I have just put it back.

File: setup/src/Magento/Setup/Model/ConfigOptionsList/DriverOptions.php (hunk 4 above)
Comment (bnymn) [reply setup/src/Magento/Setup/Model/ConfigOptionsList/DriverOptions.php:43]: Is it because of the two consecutive for-each loops?

**Comment by bnymn** [comment]
Hi @sivaschenko,

I am sorry that I could not make the changes earlier. I have just sent commits for your change requests.

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-2983](https://jira.corp.magento.com/browse/ENGCOM-2983) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Comment by sidolov** [comment]
Hi @bnymn, unfortunately, we cannot deliver your feature to 2.3.3 release because provided solution caused several issues on the other Magento services. We are going to restore your commits and re-work some places to achieve the quality goal and include changes to 2.3.4 release.
Thank you for your collaboration! I'm really sorry that we need to postpone such significant changes.

**Pull request merged**